component-base: introduce metrics api

rexagod opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
Introduce metrics API for configuring options in components in a non-breaking manner.



#### Special notes for your reviewer:

This has been branched out from https://github.com/kubernetes/kubernetes/pull/123426, as its own effort, and blocks that PR currently.

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs
The metrics (`k8s.io/component-base/metrics/`) API allows for leveraging its options in components in a non-breaking manner. Components are expected to incorporate this, over time, to offer more customizability to the users in terms of their overall observability experience on a per-component level.
```

dims commented:
does this need API review?

rexagod commented:
@dims There's some outstanding changes I've planned to get in here, will keep this thread posted once its ready for API review. Moving to draft now.

rexagod committed:
component-base: introduce metrics api

Introduce metrics API for configuring options.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod commented:
@dims Apologies for the delay here, this seems fine for API review.

Also, please note that the commit used here is cherry-picked from https://github.com/kubernetes/kubernetes/pull/123426.

rexagod commented:
Bumping!

dgrisonnet commented:
Adding the label to get some eyes on the PR from API reviewers.

/label api-review

tallclair commented on staging/src/k8s.io/component-base/metrics/options.go:
```diff
@@ -27,11 +27,22 @@ import (
+	// The key's format is <MetricName>,<LabelName>, while its value is a list of allowed values for that label of that metric. For e.g., "metric1,label1": "v1,v2,v3".
+	AllowListMapping map[string]string `json:"allowListMapping,omitempty"`
+	// The path to the manifest file that contains the allow-list mapping. Provided for convenience over AllowListMapping.
+	AllowListMappingManifest string `json:"allowListMappingManifest,omitempty"`
```
It didn't fail the test because this PR doesn't add the options to the Kubelet, that's handled in https://github.com/kubernetes/kubernetes/pull/123426, which does update the test:

https://github.com/kubernetes/kubernetes/pull/123426/files#diff-6c3cf9345a9d7b45576cdf68025b779cb844b0c05d7315b06e9b80862c2a26a6

But that is adding this to the non-path fields, which seems wrong...

liggitt replied:
> It didn't fail the test because this PR doesn't add the options to the Kubelet, that's handled in #123426, which does update the test:

oh, great :)

> But that is adding this to the non-path fields, which seems wrong...

yeah, that's definitely incorrect :)

rexagod committed:
chore: reorganize `component-base/metrics`

This commit reorganizes component-base's metrics package to bring about
the following changes:
* migrate all related functions and methods to options.go (from
  registry.go)
* inline v1.MetricsConfiguration in Options and expose it in api/v1,
  also re-generate deepcopy assets for the same
* add missing validation functions for MetricsConfiguration fields
* add missing setter functions for MetricsConfiguration fields
* addressed a logical bug in SetDisabledMetric to not increment the
  disabled metrics counter when a duplicate metric is disabled again
* refactored some tests to also take into consideration meta metrics,
  wherever applicable
* changed validateAllowListMappingManifest's godoc to mention that
  components are expected to pass in the *absolute* path to the manifest
  file

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod commented on staging/src/k8s.io/component-base/metrics/options.go:
```diff
@@ -27,11 +27,22 @@ import (
+	ShowHiddenMetricsForVersion string `json:"showHiddenMetricsForVersion,omitempty"`
+	// DisabledMetrics is a list of fully qualified metric names that should be disabled.
+	// Disabling metrics is higher in precedence than showing hidden metrics.
+	DisabledMetrics []string `json:"disabledMetrics,omitempty"`
```
Regex based validation is in [`validateDisabledMetrics`](https://github.com/kubernetes/kubernetes/pull/131572/files#diff-b71871c7a351e5b89f31ab6a34cbc68f8242f904c7640250757f213a8d4ad50aR256), the dedup is handled in the [`SetDisabledMetric`](https://github.com/kubernetes/kubernetes/pull/131572/files#diff-b71871c7a351e5b89f31ab6a34cbc68f8242f904c7640250757f213a8d4ad50aR136) function.

rexagod commented on staging/src/k8s.io/component-base/metrics/options.go:
```diff
@@ -27,11 +27,22 @@ import (
 
 // Options has all parameters needed for exposing metrics from components
+// +k8s:deepcopy-gen=true
 type Options struct {
```
Kubelet is expected to call [`Validate`](https://github.com/kubernetes/kubernetes/pull/131572/files#diff-b71871c7a351e5b89f31ab6a34cbc68f8242f904c7640250757f213a8d4ad50aR301) at the moment. I'm not sure what "flag-validation step" means here.

rexagod commented on staging/src/k8s.io/component-base/metrics/options.go:
```diff
@@ -120,17 +241,85 @@ func validateShowHiddenMetricsVersion(currentVersion semver.Version, targetVersi
+// TODO: This is kept here for backward compatibility in Kubelet (as metrics configuration fields were exposed on an individual basis earlier).
+// TODO: Revisit this after Kubelet supports the new metrics configuration API.
+func ValidateShowHiddenMetricsVersion(v string) []error {
+	err := validateShowHiddenMetricsVersion(parseVersion(version.Get()), v)
```
Not sure if I follow, could you elaborate the need for this change?

rexagod commented on staging/src/k8s.io/component-base/metrics/api/v1/types.go:
```diff
@@ -0,0 +1,40 @@
+	AllowListMapping map[string]string `json:"allowListMapping,omitempty"`
+	// The path to the manifest file that contains the allow-list mapping. Provided for convenience over AllowListMapping.
+	// The file contents must represent a map of string keys and values, i.e.,
+	// allowListMapping:
```
Ah, [right](https://github.com/kubernetes/kubernetes/blob/d2ef8a18089c643be55aa81fa5241826c5a2dd81/staging/src/k8s.io/component-base/metrics/opts_test.go#L150-L167)!

rexagod commented on staging/src/k8s.io/component-base/metrics/options.go:
```diff
@@ -120,17 +241,85 @@ func validateShowHiddenMetricsVersion(currentVersion semver.Version, targetVersi
+}
+
+// Validate validates metrics flags options.
+func (o *Options) Validate() []error {
```
ACK, moved `Validate` to `api/v1`.

rexagod commented:
/retest-required

liggitt commented on staging/src/k8s.io/component-base/metrics/api/v1/config.go:
```diff
@@ -0,0 +1,124 @@
+
+func validateDisabledMetrics(names []string, fldPath *field.Path) field.ErrorList {
+	errs := field.ErrorList{}
+	metricNameRegex := regexp.MustCompile(`^` + metricNameExpr + `$`)
```
move this to a package var since it's statically compiled

rexagod committed:
test: Address globals pollution in `metrics` tests

Make all globals reset states during and after tests that manipulate
them.

Signed-off-by: Pranshu Srivastava <rexagod@gmail.com>

rexagod commented:
Pinging @liggitt for another LGTM here (tests are passing now).
